docs/tutorial/alf_snapshot_and_advanced_play.rst

```diff
@@ -0,0 +1,249 @@
+
+In the above training example, if we add a command-line config parameter:
+
+.. code-block:: bash
+
+    rm -rf /tmp/alf_tutorial1
+    python -m alf.bin.train --root_dir /tmp/alf_tutorial1 --conf <ALF_ROOT>/alf/examples/tutorial/minimal_example_conf.py --conf_param="TrainerConfig.summary_interval=100"
+
+Again, ALF will store a snapshot at ``/tmp/alf_tutorial1/alf`` and we can get
+the config file at
+
+.. code-block:: bash
+
+    /tmp/alf_tutorial1/alf/examples/tutorial/minimal_example_conf.py
+
+However, this is just a copy of the original config file: it *doesn't* record
+our command-line parameter ``TrainerConfig.summary_interval=100``.
+
+In contrast, if we look at ``/tmp/alf_tutorial1/alf_config.py``, we'll see something
+like
+
+.. code-block:: python
+
+    ########### pre-configs ###########
+
+    import alf
+    alf.pre_config({
+        'TrainerConfig.summary_interval': 100,
+    })
+
+    ########### end pre-configs ###########
+
+on the very top of the file.
+
+Regardless of whether having the flag ``--use_alf_snapshot`` when playing a model,
+ALF will always use ``alf_config.py``. So if we'd like to make changes to the
+config file for play, we need to modify ``alf_config.py`` in either case.
+For other changes to make for play, we need to modify the snapshot code if
+``--use_alf_snapshot`` is provided, and modify the current ALF repo otherwise.
+
+Advanced play by rendering
+--------------------------
+
+Besides playing with a snapshot, another advanced play case is to utilize the
+:mod:`.alf.summary.render` module. This module contains several helper functions
+that convert arrays and tensors to :class:`~.render.Image` objects for visualization
+on screen or in a video. Any :class:`~.render.Image` object, once put into
+``info`` of an ``AlgStep`` returned by an algorithm's ``predict_step()``, will
+be concatenated to the corresponding environment frame by the side. Below we'll
+walk through an example to show how to use :mod:`.alf.summary.render`. The complete
+code is located at :mod:`.alf.examples.tutorial.ac_render_conf`.
+
+We will again train a model on the "CartPole-v0" task. So first of all, we import
+all the configuration from :mod:`.alf.examples.ac_cart_pole_conf`:
+
+.. code-block:: python
+
+    from alf.examples import ac_cart_pole_conf
+
+And import the :mod:`~.alf.summary.render` module
+
+.. code-block:: python
+
+    import alf.summary.render as render
+
+Then to tell the :mod:`~.alf.bin.play` module what to render, we overwrite
+``predict_step()`` of the original AC algorithm:
+
+.. code-block:: python
+
+    class ACRenderAlgorithm(ActorCriticAlgorithm):
+       def predict_step(self, inputs, state):
+           alg_step = super().predict_step(inputs, state)
+           action = alg_step.output
+           action_dist = alg_step.info.action_distribution
+           with alf.summary.scope("ACRender"):
+               # Render an action image
+               action_img = render.render_action(
```

Comment:
It would be great if we can explicitly label the type of the `action_img` here, so that the reader knows `render_action` returns an image. I think you mentioned that above.

Reply:
How do I explicitly label the type?

Reply:
> It would be great if we can explicitly label the type of the `action_img` here, so that the reader knows `render_action` returns an image. I think you mentioned that above.

How do I explicitly label the type?

Reply:
I was thinking about `action_img: render.Image = ...`, but comment works too.